### ISSUE TITLE
[build] Instancing functions entry points not defined on RPI

### ISSUE THREAD

**Issue by purewack** [raysan5/raylib#1679]
I am trying to build the library on a fresh raspberry pi zero with all needed things already installed like build-essentials. 

There are errors possibly related to the new issue of the rlgl redesign, but the following output of make might be useful:

```
pi@raspberrypi:~/raylib/src $ make PLATFORM=PLATFORM_RPI
gcc -c core.c -Wall -D_DEFAULT_SOURCE -Wno-missing-braces -Werror=pointer-arith -fno-strict-aliasing -std=c99 -fPIC -I. -Iexternal/glfw/include -Iexternal/glfw/deps/mingw -I/opt/vc/include -I/opt/vc/include/interface/vmcs_host/linux -I/opt/vc/include/interface/vcos/pthreads -DPLATFORM_RPI -DGRAPHICS_API_OPENGL_ES2
In file included from core.c:136:
rlgl.h:888:8: error: unknown type name ‘PFNGLDRAWARRAYSINSTANCEDEXTPROC’
 static PFNGLDRAWARRAYSINSTANCEDEXTPROC glDrawArraysInstanced = NULL;
        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
rlgl.h:888:64: warning: initialization of ‘int’ from ‘void *’ makes integer from pointer without a cast [-Wint-conversion]
 static PFNGLDRAWARRAYSINSTANCEDEXTPROC glDrawArraysInstanced = NULL;
                                                                ^~~~
rlgl.h:889:8: error: unknown type name ‘PFNGLDRAWELEMENTSINSTANCEDEXTPROC’
 static PFNGLDRAWELEMENTSINSTANCEDEXTPROC glDrawElementsInstanced = NULL;
        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
rlgl.h:889:68: warning: initialization of ‘int’ from ‘void *’ makes integer from pointer without a cast [-Wint-conversion]
 static PFNGLDRAWELEMENTSINSTANCEDEXTPROC glDrawElementsInstanced = NULL;
                                                                    ^~~~
rlgl.h:890:8: error: unknown type name ‘PFNGLVERTEXATTRIBDIVISOREXTPROC’
 static PFNGLVERTEXATTRIBDIVISOREXTPROC glVertexAttribDivisor = NULL;
        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
rlgl.h:890:64: warning: initialization of ‘int’ from ‘void *’ makes integer from pointer without a cast [-Wint-conversion]
 static PFNGLVERTEXATTRIBDIVISOREXTPROC glVertexAttribDivisor = NULL;
                                                                ^~~~
In file included from core.c:136:
rlgl.h: In function ‘rlglInit’:
rlgl.h:1709:38: error: ‘PFNGLDRAWARRAYSINSTANCEDEXTPROC’ undeclared (first use in this function); did you mean ‘PFNGLMULTIDRAWARRAYSEXTPROC’?
             glDrawArraysInstanced = (PFNGLDRAWARRAYSINSTANCEDEXTPROC)eglGetProcAddress("glDrawArraysInstancedANGLE");
                                      ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
                                      PFNGLMULTIDRAWARRAYSEXTPROC
rlgl.h:1709:38: note: each undeclared identifier is reported only once for each function it appears in
rlgl.h:1709:70: error: expected ‘;’ before ‘eglGetProcAddress’
            glDrawArraysInstanced = (PFNGLDRAWARRAYSINSTANCEDEXTPROC)eglGetProcAddress("glDrawArraysInstancedANGLE");
                                                                     ^~~~~~~~~~~~~~~~~
                                                                     ;
rlgl.h:1710:40: error: ‘PFNGLDRAWELEMENTSINSTANCEDEXTPROC’ undeclared (first use in this function); did you mean ‘PFNGLMULTIDRAWELEMENTSEXTPROC’?
             glDrawElementsInstanced = (PFNGLDRAWELEMENTSINSTANCEDEXTPROC)eglGetProcAddress("glDrawElementsInstancedANGLE");
                                        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
                                        PFNGLMULTIDRAWELEMENTSEXTPROC
rlgl.h:1710:74: error: expected ‘;’ before ‘eglGetProcAddress’
        glDrawElementsInstanced = (PFNGLDRAWELEMENTSINSTANCEDEXTPROC)eglGetProcAddress("glDrawElementsInstancedANGLE");
                                                                     ^~~~~~~~~~~~~~~~~
                                                                     ;
rlgl.h:1711:38: error: ‘PFNGLVERTEXATTRIBDIVISOREXTPROC’ undeclared (first use in this function); did you mean ‘PFNGLISVERTEXARRAYOESPROC’?
             glVertexAttribDivisor = (PFNGLVERTEXATTRIBDIVISOREXTPROC)eglGetProcAddress("glVertexAttribDivisorANGLE");
                                      ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
                                      PFNGLISVERTEXARRAYOESPROC
rlgl.h:1711:70: error: expected ‘;’ before ‘eglGetProcAddress’
            glVertexAttribDivisor = (PFNGLVERTEXATTRIBDIVISOREXTPROC)eglGetProcAddress("glVertexAttribDivisorANGLE");
                                                                     ^~~~~~~~~~~~~~~~~
                                                                     ;
rlgl.h:1713:40: warning: comparison between pointer and integer
             if ((glDrawArraysInstanced != NULL) && (glDrawElementsInstanced != NULL) && (glVertexAttribDivisor != NULL)) RLGL.ExtSupported.instancing = true;
                                        ^~
rlgl.h:1713:77: warning: comparison between pointer and integer
     if ((glDrawArraysInstanced != NULL) && (glDrawElementsInstanced != NULL) && (glVertexAttribDivisor != NULL)) RLGL.ExtSupported.instancing = true;
                                                                     ^~

rlgl.h:1713:112: warning: comparison between pointer and integer
 ULL) && (glDrawElementsInstanced != NULL) && (glVertexAttribDivisor != NULL)) RLGL.ExtSupported.instancing = true;
                                                                     ^~

rlgl.h: In function ‘rlDrawVertexArrayElements’:
rlgl.h:3118:67: error: pointer of type ‘void *’ used in arithmetic [-Werror=pointer-arith]
     glDrawElements(GL_TRIANGLES, count, GL_UNSIGNED_SHORT, buffer + offset);
                                                                   ^
cc1: some warnings being treated as errors
make: *** [Makefile:566: core.o] Error 1```

**Comment by raysan5** [comment]
Those entry points should be defined in GL headers. In any case, definition should be:
```c
#define GLAPIENTRY
typedef void (GLAPIENTRY * PFNGLDRAWARRAYSINSTANCEDEXTPROC) (GLenum mode, GLint start, GLsizei count, GLsizei primcount);
typedef void (GLAPIENTRY * PFNGLDRAWELEMENTSINSTANCEDEXTPROC) (GLenum mode, GLsizei count, GLenum type, const GLvoid *indices, GLsizei primcount);
typedef void (GLAPIENTRY * PFNGLVERTEXATTRIBDIVISORPROC) (GLuint index, GLuint divisor);
```

Not sure how it could be addressed...

**Comment by purewack** [comment]
I have now tried with all the recommended libraries for the raspberry pi os in the case of errors - as stated in the wiki article, and have enabled the GL (full KMS) driver but same exact make output :/

**Comment by raysan5** [comment]
@purewack I've been investigating this issue. It seems default RPI [OpenGL ES 2.0 libraries (extensions)](https://github.com/raspberrypi/firmware/blob/master/opt/vc/include/GLES2/gl2ext.h) do not match latest [Khronos GLES2 libraries](https://www.khronos.org/registry/OpenGL/api/GLES2/gl2ext.h). Actually RPI one dates from 2012 while Khronos one dates from 2013-2020.

I've checked that with latest Raspberry PI OS image (Lite version), also testing on RPIZero.

The mentioned include can be found in `/opt/vc/include/GLES2/gl2ext.h`, that's the default Broadcom one but if you installed Mesa OpenGL driver, you probably have another one (same name) in another directory.

Official Khronos `gl2ext.h` header defines those entry points as:
```c
#ifndef GL_EXT_draw_instanced
#define GL_EXT_draw_instanced 1
typedef void (GL_APIENTRYP PFNGLDRAWARRAYSINSTANCEDEXTPROC) (GLenum mode, GLint start, GLsizei count, GLsizei primcount);
typedef void (GL_APIENTRYP PFNGLDRAWELEMENTSINSTANCEDEXTPROC) (GLenum mode, GLsizei count, GLenum type, const void *indices, GLsizei primcount);
#ifdef GL_GLEXT_PROTOTYPES
GL_APICALL void GL_APIENTRY glDrawArraysInstancedEXT (GLenum mode, GLint start, GLsizei count, GLsizei primcount);
GL_APICALL void GL_APIENTRY glDrawElementsInstancedEXT (GLenum mode, GLsizei count, GLenum type, const void *indices, GLsizei primcount);
#endif
#endif /* GL_EXT_draw_instanced */

#ifndef GL_EXT_instanced_arrays
#define GL_EXT_instanced_arrays 1
#define GL_VERTEX_ATTRIB_ARRAY_DIVISOR_EXT 0x88FE
typedef void (GL_APIENTRYP PFNGLVERTEXATTRIBDIVISOREXTPROC) (GLuint index, GLuint divisor);
#ifdef GL_GLEXT_PROTOTYPES
GL_APICALL void GL_APIENTRY glVertexAttribDivisorEXT (GLuint index, GLuint divisor);
#endif
#endif /* GL_EXT_instanced_arrays */
```

[Here](https://www.raspberrypi.org/forums/viewtopic.php?t=240301) there is a conversation on RPI forum saying instancing could work but I'm not sure...

In any case, the function entry points definition is not defined for RPI despite it should be. Some elegant solution should be found to avoid the crash on compilation.
